core/src/main/java/tc/oc/pgm/loot/LootableMatchModule.java

```diff
@@ -0,0 +1,348 @@
+                  // Generate the loot for this Filler
+                  boolean repeat;
+                  do {
+                    filler
+                        .getLoot()
+                        .elements(opener)
+                        .forEachOrdered(
+                            item -> {
+                              if (!slots.isEmpty()) {
+                                // For each item, remove a random slot from those remaining,
+                                // apply item mods, and put it in the slot.
+                                if (immm != null) immm.applyRules(item);
+                                slots.remove(random.nextInt(slots.size())).putItem(item);
+                              }
+                            });
+                    int newValue =
+                        triggerHistory.row(this).compute(filler, (f, i) -> i == null ? 1 : ++i);
+                    repeat = filler.usesRefillTrigger() && newValue < triggeredFillers.get(filler);
+                  } while (repeat);
+                }
+              });
+    }
+  }
+
+  private class FillableInventory extends Fillable {
+    final Inventory inventory;
+
+    FillableInventory(Inventory inventory) {
+      this.inventory = inventory;
+    }
+
+    @Override
+    public int hashCode() {
+      return inventory.hashCode();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      return obj instanceof FillableInventory
+          && inventory.equals(((FillableInventory) obj).inventory);
+    }
+
+    @Override
+    Stream<Inventory> inventories() {
+      return Stream.of(inventory);
+    }
+  }
+
+  private class FillableCache extends Fillable {
+    final Cache cache;
+
+    private FillableCache(Cache cache) {
+      this.cache = cache;
+    }
+
+    @Override
+    public int hashCode() {
+      return cache.hashCode();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      return obj instanceof FillableCache && cache.equals(((FillableCache) obj).cache);
+    }
+
+    @Override
+    Stream<Inventory> inventories() {
+      return Stream.concat(
+          cache
+              .region()
+              .getChunkPositions()
+              .flatMap(cp -> Stream.of(cp.getChunk(match.getWorld()).getTileEntities()))
+              .filter(cache.region()::contains)
+              .filter(LootableMatchModule::isFillable)
+              .filter(block -> !cache.filter().query(new BlockQuery(block)).isDenied())
+              .map(block -> ((InventoryHolder) block).getInventory()),
+          cache
+              .region()
+              .getChunkPositions()
+              .flatMap(cp -> Stream.of(cp.getChunk(match.getWorld()).getEntities()))
+              .filter(cache.region()::contains)
+              .filter(LootableMatchModule::isFillable)
+              .filter(entity -> !cache.filter().query(new EntityQuery(entity)).isDenied())
+              .map(entity -> ((InventoryHolder) entity).getInventory()));
+    }
```

Comment:
```suggestion
      return cache
              .region()
              .getChunkPositions()
              .map(cp -> cp.getChunk(match.getWorld()))
              .flatMap(ch -> Stream.concat(
                  ch.getTileEntities().stream().map(BlockQuery::new),
                  ch.getEntities().stream().map(EntityQuery::new)))
              .filter(cache.region()::contains)
              .filter(LootableMatchModule::isFillable)
              .filter(q -> !cache.filter().query(q).isDenied())
              .map(q -> q.getInventory()),
    }
```

This can likely be generified in a way where BlockQuery and EntityQuery both implement some InventoryQuery so that you can perform the `getInventory` at the end, everything else just needs to support a query as parameter instead of a block-or-entity, which shouldn't be hard i assume